### PR TITLE
fix(training): wire AzureML named checkpoints output to runtime env var

### LINKED ISSUE #855: fix(training): AzureML named checkpoint output binding silently uploads nothing
## Summary

`outputs.checkpoints` AzureML binding is broken — `_sync_checkpoint_output` writes to a literal `${{outputs.checkpoints}}` directory.

The AzureML `outputs.checkpoints` `uri_folder` binding in `training/rl/workflows/azureml/train.yaml` (and its IL twin in `training/il/workflows/azureml/lerobot-train.yaml`) silently fails to receive any checkpoint files at job-end upload time. The runtime data-capability subsystem logs `WARNING ... [UriUploadSession]:.../wd/checkpoints/ is empty. Skip uploading.` Training succeeds and the model registry entry is fine because the actual checkpoints reach the workspace via MLflow's `log_artifacts`, but the named `checkpoints` output is dead.

Observed on a smoke-test run for PR #812 (tip `999c12b`) and reproducible on prior unrelated runs — this is a long-standing latent bug, not a regression.

## Root cause

`environment_variables:` entries in an AzureML `commandJob` YAML are passed through unchanged. Unlike the `command` field, AzureML **does not** substitute `${{outputs.<name>}}` template expressions in environment variable values. Inspection of `az ml job show` on any submitted run confirms the variable is sent to the container literally as the string `${{outputs.checkpoints}}`:

```json
"environment_variables": {
  "TRAINING_CHECKPOINT_OUTPUT": "${{outputs.checkpoints}}"
}
```

Meanwhile the runtime exposes the real upload directory under different env var names (visible in `system_logs/data_capability/data-capability.log`):

```
checkpoints
CHECKPOINTS
AZURE_ML_OUTPUT_checkpoints
azure_ml_output_checkpoints
AZURE_ML_OUTPUT_CHECKPOINTS
```

→ all five pointing at `/mnt/azureml/cr/j/<jobid>/cap/data-capability/wd/checkpoints`.

`_sync_checkpoint_output` in `training/rl/scripts/skrl_training.py` consequently does the wrong thing:

```python
target = os.environ.get("TRAINING_CHECKPOINT_OUTPUT")  # "${{outputs.checkpoints}}"
if not target or not checkpoint_dir.exists():
    return
destination = Path(target)        # Path("${{outputs.checkpoints}}") — relative path!
...
shutil.copytree(checkpoint_dir, destination, dirs_exist_ok=True)
```

It creates a literal `./${{outputs.checkpoints}}/` directory inside the container's working directory and copies checkpoints there. The real `cap/data-capability/wd/checkpoints` directory remains empty, the upload step skips, and the `checkpoints` output asset is registered pointing at an empty blob path.

The same pattern is present unchanged in `training/il/lerobot/scripts/...` and `training/il/workflows/azureml/lerobot-train.yaml:67`. Per `grep`, both YAML files use `TRAINING_CHECKPOINT_OUTPUT: ${{outputs.checkpoints}}` and both training entry points read `TRAINING_CHECKPOINT_OUTPUT`.

## Why this went undetected

- MLflow `log_artifacts(checkpoint_dir, artifact_path="skrl-run/checkpoints")` (skrl_training.py:222) uploads the checkpoints to the MLflow artifact store independently of the AzureML named output.
- `_register_checkpoint_model` uses the MLflow `runs:/` URI, not the AzureML output URI, so model registration works.
- The "empty folder, skip uploading" message is logged at `WARNING`, not `ERROR` — the job stays `Completed`.
- The `Path("${{outputs.checkpoints}}")` directory created on disk is benign and discarded with the container.

## Reproduction

```bash
./training/rl/scripts/submit-azureml-training.sh \
  --compute gpu-cluster --task Isaac-Velocity-Rough-Anymal-C-v0 --max-iterations 3
# After the job completes:
az ml job download -n <run_id> -w <workspace> -g <resource_group> --all
grep "is empty. Skip uploading" artifacts/system_logs/data_capability/data-capability.log
# Output: WARNING ... [UriUploadSession]:.../wd/checkpoints/ is empty. Skip uploading.
```

The "Outputs" tab of the run in Azure ML Studio also shows an empty `checkpoints` folder.

## Decision needed: fix or delete

**Option A — fix the binding (keep `outputs.checkpoints` working).**

Read the runtime-substituted env var instead of the template string. The cleanest version:

```python
target = (
    os.environ.get("TRAINING_CHECKPOINT_OUTPUT")
    or os.environ.get("AZURE_ML_OUTPUT_checkpoints")
)
if target and target.startswith("${{"):
    target = os.environ.get("AZURE_ML_OUTPUT_checkpoints")
```

Or — preferably — change the YAML and the script to a single source of truth: drop the broken `TRAINING_CHECKPOINT_OUTPUT` env var and have the script read `AZURE_ML_OUTPUT_checkpoints` directly (gated by `is not None`). Mirror the change in both `train.yaml` and `lerobot-train.yaml`, and in both `skrl_training.py` and the LeRobot training entry point.

**Option B — delete the dead binding.**

The MLflow artifact store is already authoritative for checkpoints (model registry URIs all use `runs:/...`). The `outputs.checkpoints` `uri_folder` adds nothing today; removing it would simplify both YAMLs and drop `_sync_checkpoint_output` plus its 5 unit tests.

**Recommendation: Option A.** The AzureML `uri_folder` output enables downstream pipeline composition (e.g., a separate evaluation job consuming `checkpoints` as a typed input via `--inputs checkpoints=azureml:job:<train_id>:checkpoints`). MLflow `runs:/` URIs require an MLflow client and don't compose into AzureML job graphs as cleanly. Keep the typed output, fix the wiring.

## Scope

Fix lands in:
- `training/rl/workflows/azureml/train.yaml`
- `training/rl/scripts/skrl_training.py` (`_sync_checkpoint_output`)
- `training/il/workflows/azureml/lerobot-train.yaml`
- `training/il/lerobot/scripts/` (LeRobot training entry — same env-var read pattern)
- `training/tests/test_skrl_training.py` (5 `_sync_checkpoint_output` tests already cover env-var behaviour; they monkeypatch `TRAINING_CHECKPOINT_OUTPUT` directly so the assertions are correct, but the var name should change with the fix)

## Verification

After the fix, submit any short training run and confirm the data-capability log shows an upload (e.g., `[UriUploadSession]:.../wd/checkpoints/ uploaded N files`) instead of the empty-folder skip warning. Verify the run's "Outputs" tab in Studio shows `.pt` files under `checkpoints/`.

## References

- Related convention: AzureML inputs with `mode=download` expose paths via env var `AZURE_ML_INPUT_<input_name>`; the `${{inputs.X}}` command template ref is NOT substituted for downloaded inputs, only for `ro_mount`. The same constraint applies to outputs in `environment_variables`.
- Affected files: `training/rl/workflows/azureml/train.yaml:58`, `training/rl/scripts/skrl_training.py:124-138`, `training/il/workflows/azureml/lerobot-train.yaml:67`

### PR DESCRIPTION
**Pull request by algattik**
# Pull Request

## Description

Fixes the AzureML `checkpoints` *uri_folder* output for both RL and IL training jobs (issue #855), then adds the surrounding pieces required for that fix to be exercisable end-to-end on a scale-to-zero GPU cluster.

The root cause was wiring the named output through `TRAINING_CHECKPOINT_OUTPUT: ${{outputs.checkpoints}}` inside `environment_variables:`. Azure ML only substitutes `${{outputs.X}}` template expressions inside `command:`; inside `environment_variables:` the container receives the literal template string, writes into a directory of that name, and the data-capability sidecar uploads nothing to the named output. Training entry points now read `$AZURE_ML_OUTPUT_CHECKPOINTS` directly (the per-output runtime path Azure ML exports for every `uri_folder` output) and mirror their local *checkpoints/* tree into it at finalization. The broken `${{outputs.checkpoints}}` line is removed from both job YAMLs.

A real AzureML LeRobot e2e test was added alongside the existing SKRL one, and `assert_job_has_checkpoint` was upgraded to enumerate the *checkpoints* output blobs via the **azure-storage-blob SDK** and assert non-empty contents — the previous declaration-only check happily passed against the buggy code. The LeRobot test forced two related issues to the surface and they are fixed in the same branch: a *torchcodec* / *torch* C++ ABI mismatch from a stale override pin, and missing MLflow metrics on short runs because lerobot's `cfg.log_freq` defaults to 200. Finally, three independent defaults blocked scale-from-zero on the GPU pool the new test targets — *aml-operator* resource validation, a missing static `accelerator=nvidia` node label, and Volcano's enqueue-time capacity gate — and all three are corrected and documented.

Closes #855

### Checkpoint output wiring

- Replaced the `TRAINING_CHECKPOINT_OUTPUT` env-var indirection with a direct read of **`$AZURE_ML_OUTPUT_CHECKPOINTS`** in *training/rl/scripts/skrl_training.py* and a new `_sync_checkpoint_output` helper in *training/il/scripts/lerobot/train.py*. Both entry points mirror local *checkpoints/* into the runtime path on the SIGTERM handler and on normal finalization.
- Removed the literal `TRAINING_CHECKPOINT_OUTPUT: ${{outputs.checkpoints}}` line from *training/rl/workflows/azureml/train.yaml* and *training/il/workflows/azureml/lerobot-train.yaml*; both YAMLs now carry an inline comment warning that `environment_variables:` does **not** template-substitute.
- Updated *training/specifications/rl-training.specification.md*, *docs/operations/troubleshooting.md*, and *.github/copilot-instructions.md* to describe the new contract.

### AzureML e2e coverage

- Rewrote `assert_job_has_checkpoint` in *tests/e2e/_aml.py* to resolve the `workspaceblobstore` datastore via `MLClient.datastores.get(...)`, then list blobs under `azureml/{job}/checkpoints/` with `azure.storage.blob.ContainerClient.list_blobs(name_starts_with=...)`. The check no longer downloads any checkpoint bytes — avoids the >100 MB AzCopy nag on every e2e run — but still catches the exact regression #855 fixed; the previous payload-only check did not.
- Added `submit_aml_lerobot_training` and `test_aml_il_training_e2e` (lerobot/pusht, *act*, 10 steps, `--log-freq 1`, `--eval-freq > training-steps` to skip in-loop eval).
- Parametrised `_assert_run_has_expected_tracking` in *tests/e2e/_mlflow.py* with `required_metrics` / `required_params` and added a LeRobot-specific tuple (`train/loss`, `train/learning_rate`, `policy_type`, `num_gpus`, `distributed`) plus `assert_aml_lerobot_job_has_mlflow_tracking`.
- Added `TestSyncCheckpointOutput` to *training/tests/test_lerobot_train.py* (five cases mirroring the SKRL coverage); renamed the env var in the matching SKRL tests.

### LeRobot dependency fixes

- Pinned **`torchcodec==0.10.0`** (was 0.13.0) in *training/il/lerobot/pyproject.toml*. 0.13.x requires `torch>=2.11` per its release notes but declares no torch dependency in PyPI metadata, so the uv resolver did not catch it — pairing 0.13.0 with the lerobot-allowed torch 2.10 produced `libtorchcodec_core4.so: undefined symbol: torch_from_blob` at import time. 0.10.0 still ships `manylinux_2_28_x86_64` wheels for cp312.
- Pinned **`numpy==2.2.6`** (was 2.4.6) — the previous comment claimed lerobot's cap was `<2.4` but the actual cap on 0.5.1 is `<2.3`.
- Dropped both pins from `[tool.uv] override-dependencies` and regenerated *training/il/lerobot/requirements.txt*.

### MLflow `--log-freq` plumbing

- lerobot emits the `step:N smpl:... loss:... lr:...` lines our wrapper parses only when `step % cfg.log_freq == 0`; with the upstream default of 200 the 10-step e2e run captured zero MLflow metrics. Added `--log-freq N` end-to-end: *submit-azureml-lerobot-training.sh* CLI flag and `LOG_FREQ` env-var passthrough, `"LOG_FREQ": "--log_freq"` in `env_arg_map` inside *training/il/scripts/lerobot/train.py*, the new required `log_freq` kwarg on `submit_aml_lerobot_training`, and `log_freq=1` in the e2e test.

### Scale-from-zero GPU pools

> Three independent defaults deadlock scale-from-zero on the AzureML GPU pool. *aml-operator* rejects any job whose `InstanceType` exceeds the largest currently-Ready node (which is "none" on an empty pool, so submission fails before a Pod exists for the autoscaler to react to). The autoscaler's synthetic node template strips the `accelerator: nvidia` label that the InstanceType selector requires. And Volcano's `enqueue` action runs `JobEnqueueable` plugins (`overcommit`, `proportion`) that compare requested resources against currently-Ready cluster capacity — when GPU total is zero, every GPU PodGroup is rejected at enqueue, no Pod is ever created, and the autoscaler never sees a Pending Pod.

- Flipped `amloperator.skipResourceValidation` to **`true`** by default in *infrastructure/setup/02-deploy-azureml-extension.sh* and the matching `azureml-aks-config.template.json`. New `--enforce-resource-validation` flag opts back into fail-fast for fixed-capacity clusters.
- Added the static **`node_labels = { accelerator = "nvidia" }`** declaration to the default GPU pool in *infrastructure/terraform/variables.tf* and *infrastructure/terraform/modules/sil/variables.tf* so the autoscaler's synthetic node template includes it.
- Patched the `volcano-scheduler-configmap` post-install to drop **`overcommit`** and **`proportion`** from tier 3 — gang scheduling stays in tier 2 so multi-pod jobs still wait for `minAvailable` before any task binds. New manifest *infrastructure/setup/manifests/volcano-scheduler-config-scale-from-zero.conf* and `--enforce-volcano-capacity-check` flag in *02-deploy-azureml-extension.sh* opt back into the upstream behaviour on multi-tenant clusters where submit-time queue fairness must be preserved.
- Documented all three root causes, the override flags, the static-label requirement, and a three-layer troubleshooting walk in the *🛌 Scale-from-zero GPU Pools* section of *docs/training/azureml-training.md*. Cross-linked from *docs/infrastructure/manage-node-pools.md*. Added `amloperator`, `enqueueable`, `inqueue`, and `podgroup` to *.cspell/general-technical.txt*.

## Type of Change

- [x] 🐛 Bug fix (non-breaking change fixing an issue)
- [ ] ✨ New feature (non-breaking change adding functionality)
- [ ] 💥 Breaking change (fix or feature causing existing functionality to change)
- [x] 📚 Documentation update
- [x] 🏗️ Infrastructure change (Terraform/IaC)
- [ ] ♻️ Refactoring (no functional changes)

## Component(s) Affected

- [ ] `infrastructure/terraform/prerequisites/` - Azure subscription setup
- [x] `infrastructure/terraform/` - Terraform infrastructure
- [x] `infrastructure/setup/` - OSMO control plane / Helm
- [ ] `workflows/` - Training and evaluation workflows
- [x] `training/` - Training pipelines and scripts
- [x] `docs/` - Documentation

## Testing Performed

- [ ] Terraform `plan` reviewed (no unexpected changes)
- [ ] Terraform `apply` tested in dev environment
- [ ] Training scripts tested locally with Isaac Sim
- [ ] OSMO workflow submitted successfully
- [ ] Smoke tests passed (`smoke_test_azure.py`)

End-to-end verification on `mlw-hexmshack-dev-001` (AKS `aks-hexmshack-dev-001`, GPU pool `a100gpu` at `count=0, max=1`, Standard_NC24ads_A100_v4):

- `uv run pytest -vv -s -m e2e tests/e2e/test_e2e_training.py::test_aml_il_training_e2e` — **passed** (status *Completed*, ~6 min wall time). Verified the full chain: submit → queue → autoscale GPU pool 0→1 → Pod binds → run → MLflow tracking → non-empty `checkpoints` named output enumerated via blob list. MLflow captured `train/loss`, `train/learning_rate`, `policy_type=act`, `num_gpus=1`, `distributed=false`.
- Unit suites covering `_sync_checkpoint_output` (SKRL + LeRobot) extended with cases for the new `AZURE_ML_OUTPUT_CHECKPOINTS` env var.

## Documentation Impact

- [ ] No documentation changes needed
- [x] Documentation updated in this PR
- [ ] Documentation issue filed

## Bug Fix Checklist

- [x] Linked to issue being fixed
- [x] Regression test included, OR
- [ ] Justification for no regression test:

## Checklist

- [x] My code follows the [project conventions](copilot-instructions.md)
- [x] Commit messages follow [conventional commit format](instructions/commit-message.instructions.md)
- [x] I have performed a self-review
- [x] Documentation impact assessed above
- [x] No new linting warnings introduced

### CODE REVIEW

**Review by bindsi — APPROVED** [review]
Automated review: Well-constructed fix for AzureML checkpoint output wiring. The key insight — that Azure ML doesn't substitute `${{outputs.X}}` template expressions inside `environment_variables` (only inside `command`) — is correctly addressed by reading the runtime-injected `AZURE_ML_OUTPUT_CHECKPOINTS` env var directly. The e2e test improvement (verifying actual blob contents rather than just output declaration existence) catches exactly the silent-empty-upload bug described in #855. torchcodec version alignment (0.13→0.10) with proper justification for the torch/lerobot compatibility constraint is sound. LGTM.

**Review by rezatnoMsirhC — APPROVED** [review]
